### ISSUE TITLE
Session timeout handling on JSF AJAX request

### ISSUE THREAD

**Issue by eduarddrenth** [primefaces/primefaces#4109]
Lot's of discussions and questions about this on the internet.

## 1) Environment
- PrimeFaces version: 6.2(.9)
- Does it work on the newest released PrimeFaces version? No
- Does it work on the newest sources in GitHub? No
- Application server + version: payara 5.183
- Affected browsers: all
- PrimeExceptionHandlerFactory and ajaxExceptionHandler in place

## 2) Expected behavior

when an ajax request fires and my session is invalid I want to know and I want to get a chance to login again

## 3) Actual behavior

when an ajax request fires and my session is invalid I see nothing, under the hood I get a response that looks like this:

```<?xml version='1.0' encoding='UTF-8'?>
<partial-response><changes><update id="j_id1:javax.faces.ViewState:0"><![CDATA[3793564356142802147:-5855215245280144347]]></update></changes></partial-response>
```

## 4) Steps to reproduce

develop an app with security context and login with at least one protected page that fires ajax

## 5) Sample XHTML

left as a todo

## 6) Sample bean

can be any simple bean

This issue is solved by PR #4103

**Comment by tandraschko** [comment]
You should actually get a ViewExpiredException, maybe a bug in the JSF impl you use.
And for ViewExpiredException, we have a ExceptionHandler. So it's probably a won't fix.

**Comment by eduarddrenth** [comment]
Perhaps you are right, but I have everything in place, PrimeExceptionHandlerFactory and ajaxExceptionHandler, error pages, security setup. Payara is very standard and robust and aimed at production use, so I don't expect issues there. I always develop stuff as default and standard as I can, so there's nothing goofy in my code or setup. I will do some more testing and get back, but I'm pretty sure there is an issue here and the listener solves it.

**Comment by eduarddrenth** [comment]
If I understand correctly you never see an exception becasue your EE server is doing a forward to your login page. So, to deal with this in a friendly way, from my javadoc:

```
This is how it works:

    ajax request encounters invalid session
    this listener detects DispatcherType.FORWARD to getLoginPath()
    handleAjaxAfterTimeout(javax.faces.context.FacesContext, javax.servlet.http.HttpServletRequest) is called
    In this method a redirect is done to getTimeoutPath(javax.servlet.http.HttpServletRequest), which should be a protected page, possibly via a non-protected page using meta refresh and showing some friendly message
    your security setup redirects to the login page
    after login you land on the page from step 4

declare this phase-listener in your faces-config
```

**Comment by tandraschko** [comment]
I only know the default case: when you do a post, with a expired viewId, a viewExpired is thrown. If this doesn't happen, something is wrong.

**Comment by eduarddrenth** [comment]
So what you are saying is that payara/glassfish is handling this wrongly? Can it be a spec change? I use EE 8, cdi 2, jsf 2.3, servlet 4.

**Comment by tandraschko** [comment]
Probably yes.
Just give MyFaces and Mojarra a try on a simple Tomcat and see what happens.

**Comment by eduarddrenth** [comment]
ok, I will

**Comment by eduarddrenth** [comment]
@tandraschko I'd like to get back to this quote:

> I only know the default case: when you do a post, with a expired viewId, a viewExpired is thrown. If this doesn't happen, something is wrong.

A post is done, but the session is timed out, the post with the viewid never reaches the servlet. The post will be forwarded to the login page by the server, which is pretty default behaviour isn't it? So a ViewExpired is not thrown. Forwarding to a login page in case of ajax is not usefull. Actually a redirect would be more helpfull, hence the listener. Only after you have a new session you may encounter ViewExpired, depending on wether you repost or not. Only than the exceptionhandler can do it's job.

Testing with myfaces/mojarra/tomcat is a bit cumbersum for me now, but I'll see if I can find the time somehow.

**Comment by eduarddrenth** [comment]
Long ago this one, this is still an issue and I still have a solution for it working nicely in several production apps.

**Comment by melloware** [comment]
I am with @tandraschko on this. It feels wrong to include this with Primefaces.

However maybe we can put it in Primefaces Extensions if you want?

And since its optional and will only be turned on if someone includes it in their faces-config.xml it should be fine to include.
